gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala

```diff
@@ -219,27 +219,29 @@ trait HashJoinLikeExecTransformer
 
   override def doTransform(substraitContext: SubstraitContext): TransformContext = {
 
-    def transformAndGetOutput(plan: SparkPlan): (RelNode, Seq[Attribute], Boolean) = {
+    def transformAndGetOutput(plan: SparkPlan): (RelNode, Seq[Attribute], Boolean, JLong) = {
       plan match {
         case p: TransformSupport =>
           val transformContext = p.doTransform(substraitContext)
-          (transformContext.root, transformContext.outputAttributes, false)
+          (transformContext.root, transformContext.outputAttributes, false, -1L)
         case _ =>
           val readRel = RelBuilder.makeReadRel(
             new JArrayList[Attribute](plan.output.asJava),
             substraitContext,
             -1
           ) /* A special handling in Join to delay the rel registration. */
-          (readRel, plan.output, true)
+          // Make sure create a new read relId for the stream side first
+          // before the one of the build side, when there is not shuffle on the build side
+          (readRel, plan.output, true, substraitContext.nextRelId())
```

Comment:
@rui-mo  @PHILO-HE  please help to check, this change will make sure that it creates a new relId for the stream side before the build side, when the stream side is reading data from iterator and the build side is no shuffle.

Reply:
When there is a shuffle in the stream side and no shuffle in the build side, the operation orders passed to the backend are: 0: `stream side read from iter`, 1:`build side scan`, 2: `build side filter`, 4: `join`, but the orders created in the join transformer are: 0:`build side scan`, 1: `build side filter`, 2: `stream side read from iter`, 4: `join`, because `stream side read from iter` will be registed after creating the build side, so they are different.